README.md
# Correios CEP

[![Hex.pm](https://img.shields.io/hexpm/v/correios_cep.svg)](https://hex.pm/packages/correios_cep)
[![Docs](https://img.shields.io/badge/hex-docs-542581.svg)](https://hexdocs.pm/correios_cep)
[![Build Status](https://travis-ci.org/prodis/correios-cep-elixir.svg?branch=master)](https://travis-ci.org/prodis/correios-cep-elixir)
[![Coverage Status](https://coveralls.io/repos/github/prodis/correios-cep-elixir/badge.svg?branch=master)](https://coveralls.io/github/prodis/correios-cep-elixir?branch=master)
[![License](https://img.shields.io/hexpm/l/correios_cep.svg)](LICENSE)

![Correios Logo](http://prodis.net.br/images/ruby/2015/correios_logo.png)

Find Brazilian addresses by postal code, directly from Correios API. No HTML parsers.

## DEPRECATION NOTE

`Correios.CEP.Address.zipcode` was removed in the version `0.6.0`. Use `Correios.CEP.Address.postal_code` instead.

## Installation

The package can be installed by adding `correios_cep` to your list of dependencies in `mix.exs`:

```elixir
def deps do
  [
    {:correios_cep, "~> 0.7.0"}
  ]
end
```

## Usage

```elixir
iex> Correios.CEP.find_address("54250610")
{:ok,
 %Correios.CEP.Address{
   street: "Rua Fernando Amorim",
   neighborhood: "Cavaleiro",
   complement: "",
   city: "Jaboatão dos Guararapes",
   state: "PE",
   postal_code: "54250610"
 }}

iex> Correios.CEP.find_address("00000-000")
{:error,
 %Correios.CEP.Error{
   type: :postal_code_not_found,
   message: "Postal code not found",
   reason: "CEP NAO ENCONTRADO"
 }}

iex> Correios.CEP.find_address!("54250-610")
%Correios.CEP.Address{
  street: "Rua Fernando Amorim",
  neighborhood: "Cavaleiro",
  complement: "",
  city: "Jaboatão dos Guararapes",
  state: "PE",
  postal_code: "54250610"
}

iex> Correios.CEP.find_address!("00000-000")
** (Correios.CEP.Error) Postal code not found
```

### Options

There are some supported options that can be added to the request, as timeouts, proxy and URL
configuration.

The example below shows the use of `request_timeout` and `proxy` options:

```elixir
iex> Correios.CEP.find_address("54250610", request_timeout: 3000, proxy: {"localhost", 8888})
{:ok,
 %Correios.CEP.Address{
   street: "Rua Fernando Amorim",
   neighborhood: "Cavaleiro",
   complement: "",
   city: "Jaboatão dos Guararapes",
   state: "PE",
   postal_code: "54250610"
 }}
```

See [`Correios.CEP.find_address/2`](https://hexdocs.pm/correios_cep/Correios.CEP.html#find_address/2)
documentation to check the details of the available options.

## Documentation

The full documentation is available at [https://hexdocs.pm/correios_cep](https://hexdocs.pm/correios_cep).

## Contributing

See the [contributing guide](https://github.com/prodis/correios-cep-elixir/blob/master/CONTRIBUTING.md).

## License

Correios CEP is released under the Apache 2.0 License. See the [LICENSE](https://github.com/prodis/correios-cep-elixir/blob/master/LICENSE) file.

Copyright © 2018-2020 Fernando Hamasaki de Amorim

## Author

[Fernando Hamasaki de Amorim (prodis)](https://github.com/prodis)

![Prodis Logo](https://camo.githubusercontent.com/c01a3ebca1c000d7586a998bb07316c8cb784ce5/687474703a2f2f70726f6469732e6e65742e62722f696d616765732f70726f6469735f3135302e676966)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [Unreleased]

## [0.7.0] - 2020-07-25
## Changed
- Improve error responses. [Issue #18](https://github.com/prodis/correios-cep-elixir/issues/18)

## [0.6.0] - 2020-06-17
## Changed
- Update HTTPoison to 1.17.0 (hackeny 1.16.0)

## Deprecated
- The field `zipcode` in `Correios.CEP.Address`. Use the field `postal_code` instead.

## Removed
- Support for OTP 20.

## [0.5.1] - 2020-06-09
### Added
- The field `postal_code` in `Correios.CEP.Address` with the same value of `zipcode` field, that
will be deprecated in the next version. Part of the [Issue #17](https://github.com/prodis/correios-cep-elixir/issues/17).

## [0.5.0] - 2020-05-24
### Added
- Support for proxy configuration. [Issue #7](https://github.com/prodis/correios-cep-elixir/issues/7)

### Changed
- Update dependencies.

## [0.4.0] - 2020-04-19
### Added
- Correios API full URL in options.
- Missing code documentation and typespecs.
- Makefile.

### Changed
- General update and refactoring.

### Fixed
- Remove dependency from config/config.exs. [Issue #15](https://github.com/prodis/correios-cep-elixir/issues/15)

## [0.3.0] - 2010-01-01
### Added
- Validate zip code. [Issue #11](https://github.com/prodis/correios-cep-elixir/issues/11)
- Typespecs and Dialyzer. [Issue #1](https://github.com/prodis/correios-cep-elixir/issues/1)

## [0.2.0] - 2018-11-26
### Added
- Timeout options. [Issue #6](https://github.com/prodis/correios-cep-elixir/issues/6)

## [0.1.0] - 2018-11-12
### Added
- Initial release.

[Unreleased]: https://github.com/prodis/correios-cep-elixir/compare/0.7.0...master
[0.7.0]: https://github.com/prodis/correios-cep-elixir/compare/0.6.0...0.7.0
[0.6.0]: https://github.com/prodis/correios-cep-elixir/compare/0.5.1...0.6.0
[0.5.1]: https://github.com/prodis/correios-cep-elixir/compare/0.5.0...0.5.1
[0.5.0]: https://github.com/prodis/correios-cep-elixir/compare/0.4.0...0.5.0
[0.4.0]: https://github.com/prodis/correios-cep-elixir/compare/v0.3.0...0.4.0
[0.3.0]: https://github.com/prodis/correios-cep-elixir/compare/v0.2.0...v0.3.0
[0.2.0]: https://github.com/prodis/correios-cep-elixir/compare/v0.1.0...v0.2.0
[0.1.0]: https://github.com/prodis/correios-cep-elixir/compare/69f7fa4...v0.1.0

The format of this file is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and
this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).


lib/correios/cep.ex
defmodule Correios.CEP do
  @moduledoc """
  Find Brazilian addresses by postal code, directly from Correios API. No HTML parsers.
  """

  alias Correios.CEP.{Address, Client, Error, Parser}

  @type t :: {:ok, Address.t()} | {:error, Error.t()}

  @postal_code_regex ~r/^\d{5}-?\d{3}$/

  @doc """
  Finds address by the given `postal_code`.

  Postal codes with and without "-" separator are accepted.

  ## Options

    * `connection_timeout`: timeout for establishing a connection, in milliseconds. Default is 5000.
    * `request_timeout`: timeout for receiving the HTTP response, in milliseconds. Default is 5000.
    * `proxy`: proxy to be used for the request: `{host, port}` tuple, where `port` is an integer.
    * `proxy_auth`: proxy authentication: `{user, password}` tuple.
    * `url`: Correios API full URL. Default is `https://apps.correios.com.br/SigepMasterJPA/AtendeClienteService/AtendeCliente`.

  ## Examples

      iex> #{inspect(__MODULE__)}.find_address("54250610")
      {:ok,
       %#{inspect(Address)}{
         city: "Jaboatão dos Guararapes",
         complement: "",
         neighborhood: "Cavaleiro",
         state: "PE",
         street: "Rua Fernando Amorim",
         postal_code: "54250610"
       }}

      iex> #{inspect(__MODULE__)}.find_address("54250-610")
      {:ok,
       %#{inspect(Address)}{
         city: "Jaboatão dos Guararapes",
         complement: "",
         neighborhood: "Cavaleiro",
         state: "PE",
         street: "Rua Fernando Amorim",
         postal_code: "54250610"
       }}

      iex> #{inspect(__MODULE__)}.find_address("54250-610", connection_timeout: 1000, request_timeout: 1000)
      {:ok,
       %#{inspect(Address)}{
         city: "Jaboatão dos Guararapes",
         complement: "",
         neighborhood: "Cavaleiro",
         state: "PE",
         street: "Rua Fernando Amorim",
         postal_code: "54250610"
       }}

      iex> #{inspect(__MODULE__)}.find_address("54250-610", proxy: {"localhost", 8888})
      {:ok,
       %#{inspect(Address)}{
         city: "Jaboatão dos Guararapes",
         complement: "",
         neighborhood: "Cavaleiro",
         state: "PE",
         street: "Rua Fernando Amorim",
         postal_code: "54250610"
       }}

      iex> #{inspect(__MODULE__)}.find_address(
      ...>   "54250-610",
      ...>   proxy: {"localhost", 8888},
      ...>   proxy_auth: {"myuser", "mypass"}
      ...> )
      {:ok,
       %#{inspect(Address)}{
         city: "Jaboatão dos Guararapes",
         complement: "",
         neighborhood: "Cavaleiro",
         state: "PE",
         street: "Rua Fernando Amorim",
         postal_code: "54250610"
       }}

      iex> #{inspect(__MODULE__)}.find_address("00000-000")
      {:error,
        %#{inspect(Error)}{
          type: :postal_code_not_found,
          message: "Postal code not found",
          reason: "CEP NAO ENCONTRADO"
        }}

      iex> #{inspect(__MODULE__)}.find_address("1234567")
      {:error,
        %#{inspect(Error)}{
          type: :postal_code_invalid,
          message: "Postal code in invalid format"
        }}

      iex> #{inspect(__MODULE__)}.find_address("")
      {:error,
        %#{inspect(Error)}{
          type: :postal_code_required,
          message: "Postal code is required"
        }}

  """
  @spec find_address(String.t(), keyword()) :: t()
  def find_address(postal_code, options \\ [])

  def find_address("", _options),
    do: {:error, Error.new(:postal_code_required, "Postal code is required")}

  def find_address(postal_code, options) when is_binary(postal_code) and is_list(options) do
    if valid_postal_code?(postal_code) do
      postal_code
      |> client().request(options)
      |> parse()
    else
      {:error, Error.new(:postal_code_invalid, "Postal code in invalid format")}
    end
  end

  @spec valid_postal_code?(String.t()) :: boolean()
  defp valid_postal_code?(postal_code), do: postal_code =~ @postal_code_regex

  @spec client :: module()
  defp client, do: Application.get_env(:correios_cep, :client) || Client

  @spec parse(Client.t()) :: t()
  defp parse({:ok, response}) do
    response
    |> Parser.parse_ok()
    |> case do
      %Address{} = address -> {:ok, address}
      %Error{} = error -> {:error, error}
    end
  end

  defp parse({:error, error}), do: {:error, Parser.parse_error(error)}

  @doc """
  Finds address by a given postal code.

  Similar to `find_address/2` except it will unwrap the error tuple and raise in case of errors.

  ## Examples

      iex> #{inspect(__MODULE__)}.find_address!("54250610")
      %#{inspect(Address)}{
        city: "Jaboatão dos Guararapes",
        complement: "",
        neighborhood: "Cavaleiro",
        state: "PE",
        street: "Rua Fernando Amorim",
        postal_code: "54250610"
      }

      iex> #{inspect(__MODULE__)}.find_address!("00000-000")
      ** (#{inspect(Error)}) Postal code not found

  """
  @spec find_address!(String.t(), keyword()) :: Address.t()
  def find_address!(postal_code, options \\ [])
      when is_binary(postal_code) and is_list(options) do
    postal_code
    |> find_address(options)
    |> case do
      {:ok, response} -> response
      {:error, %Error{message: message}} -> raise Error, message
    end
  end
end
